tradeexecutor/ethereum/uniswap_v3_routing.py

```diff
@@ -0,0 +1,691 @@
+"""Route trades to different Uniswap v2 like exchanges."""
+
+import logging
+from collections import defaultdict
+from decimal import Decimal
+from typing import Dict, Set, List, Optional, Tuple
+
+from eth_typing import HexAddress, ChecksumAddress
+from tradingstrategy.chain import ChainId
+from web3 import Web3
+from web3.contract import Contract
+
+from eth_defi.abi import get_deployed_contract
+from eth_defi.gas import estimate_gas_fees
+from eth_defi.token import fetch_erc20_details
+from eth_defi.uniswap_v3.deployment import UniswapV3Deployment, fetch_deployment
+from eth_defi.uniswap_v3.swap import swap_with_slippage_protection
+from eth_defi.gas import apply_gas
+from eth_defi.hotwallet import HotWallet
+from web3.exceptions import ContractLogicError
+from web3._utils.transactions import fill_nonce
+
+from tradeexecutor.ethereum.execution import get_token_for_asset
+from tradeexecutor.ethereum.tx import TransactionBuilder
+from tradeexecutor.state.blockhain_transaction import BlockchainTransaction
+from tradeexecutor.state.identifier import TradingPairIdentifier, AssetIdentifier
+from tradeexecutor.state.trade import TradeExecution
+from tradeexecutor.strategy.routing import RoutingModel, RoutingState, CannotRouteTrade
+from tradeexecutor.strategy.trading_strategy_universe import TradingStrategyUniverse, translate_trading_pair, \
+    translate_token
+from tradingstrategy.pair import PandasPairUniverse
+
+from tradeexecutor.strategy.universe_model import StrategyExecutionUniverse
+
+
+logger = logging.getLogger(__name__)
+
+
+class OutOfBalance(Exception):
+    """Did not have enough tokens"""
+
+
+class UniswapV3RoutingState(RoutingState):
```

Comment:
Use same base class for v2 and v3 routing, so that we do not copy paste any code